Document C++ Initializers (#22704)
nim-lang/Nim: doc/manual_experimental.md

Notice when calling a constructor in the section of a global variable initialization, it will be called before `NimMain` meaning Nim is not fully initialized.

Constructor Initializer
=======================

By default Nim initializes `importcpp` types with `{}`. This can be problematic when importing
types with a deleted default constructor. In order to avoid this, one can specify default values for a constructor by specifying default values for the proc params in the `constructor` proc.

For example:

```nim

{.emit:"""/*TYPESECTION*/
struct CppStruct {
  CppStruct(int x, char* y): x(x), y(y){}
  int x;
  char* y;
};
""".}
type
  CppStruct {.importcpp, inheritable.} = object

proc constructCppStruct(a:cint = 5, b:cstring = "hello"): CppStruct {.importcpp: "CppStruct(@)", constructor.}

> 

Reviewer: Add an actual use kinda like an `assert` or `echo` or something like that,
users sometimes complain they copypaste examples and wont run or "does nothing" etc IMHO.

Notice when calling a constructor in the section of a global variable initialization, it will be called before `NimMain` meaning Nim is not fully initialized.

Constructor Initializer
=======================

By default Nim initializes `importcpp` types with `{}`. This can be problematic when importing
types with a deleted default constructor. In order to avoid this, one can specify default values for a constructor by specifying default values for the proc params in the `constructor` proc.

For example:

```nim

> {.emit:"""/*TYPESECTION*/

Reviewer 2: ```suggestion
{.emit: """/*TYPESECTION*/
```

struct CppStruct {
  CppStruct(int x, char* y): x(x), y(y){}
  int x;
  char* y;
};

> """.}

Reviewer 3: ```suggestion
{.emit:"""
/*TYPESECTION*/
struct CppStruct {
  CppStruct(int x, char* y): x(x), y(y){}
  int x;
  char* y;
};
""".}
```

type
  CppStruct {.importcpp, inheritable.} = object

> proc makeCppStruct(a:cint = 5, b:cstring = "hello"): CppStruct {.importcpp: "CppStruct(@)", constructor.}

Reviewer 2: ```suggestion
proc makeCppStruct(a: cint = 5, b:cstring = "hello"): CppStruct {.importcpp: "CppStruct(@)", constructor.}
```

> (proc (s:CppStruct) = echo "hello")(makeCppStruct()) 

Reviewer 2: ```suggestion
(proc (s: CppStruct) = echo "hello")(makeCppStruct()) 
```

> #if one remove a default value from the constructor and pass it to the call. The C++ compiler will complain.

Reviewer 2: ```suggestion
# If one removes a default value from the constructor and passes it to the call explicitly, the C++ compiler will complain.
```
